anxcloud/common_resource_tagging.go

```diff
@@ -0,0 +1,106 @@
+package anxcloud
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sort"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"go.anx.io/go-anxcloud/pkg/api"
+	corev1 "go.anx.io/go-anxcloud/pkg/apis/core/v1"
+)
+
+func withTagsAttribute(s schemaMap) schemaMap {
+	s["tags"] = &schema.Schema{
+		Type:        schema.TypeList,
+		Optional:    true,
+		Computed:    true,
+		Description: "List of tags attached to the resource.",
+		Elem: &schema.Schema{
+			Type: schema.TypeString,
+		},
+		// suppress diff when only the order has changed
+		DiffSuppressOnRefresh: true,
+		DiffSuppressFunc: func(k, oldValue, newValue string, d *schema.ResourceData) bool {
+			o, n := d.GetChange("tags")
+			oStringArray := castInterfaceArray[string](o.([]interface{}))
+			nStringArray := castInterfaceArray[string](n.([]interface{}))
+			sort.Strings(oStringArray)
+			sort.Strings(nStringArray)
+			return reflect.DeepEqual(oStringArray, nStringArray)
+		},
+	}
+	return s
+}
+
+type schemaContextFunc func(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics
+
+func ensureTagsMiddleware(pre schemaContextFunc) schemaContextFunc {
+	return func(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+		diags := pre(ctx, d, m)
+		if diags.HasError() {
+			return diags
+		}
+
+		// we don't touch remote tags when tags attribute is not set
+		// remote tags are also kept when tags attribute was unset
+		if d.GetRawConfig().GetAttr("tags").IsNull() {
+			return diags
+		}
+
+		if err := ensureTags(ctx, m.(providerContext).api, d.Id(), castInterfaceArray[string](d.Get("tags").([]interface{}))); err != nil {
+			diags = append(diags, diag.FromErr(err)...)
+		}
+
+		return diags
+	}
+}
+
+func readTagsMiddleware(pre schemaContextFunc) schemaContextFunc {
+	return func(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+		diags := pre(ctx, d, m)
+		if diags.HasError() {
+			return diags
+		}
+
+		tags, err := readTags(ctx, m.(providerContext).api, d.Id())
+		if err != nil {
+			return append(diags, diag.FromErr(err)...)
+		}
+
+		if err := d.Set("tags", tags); err != nil {
+			diags = append(diags, diag.FromErr(err)...)
+		}
+
+		return diags
+	}
+}
+
+func ensureTags(ctx context.Context, a api.API, resourceID string, tags []string) error {
+	resource := corev1.Resource{Identifier: resourceID}
+
+	remote, err := readTags(ctx, a, resourceID)
+	if err != nil {
+		return fmt.Errorf("failed to fetch remote tags: %w", err)
+	}
+
+	if toRemove := sliceSubstract(remote, tags); len(toRemove) > 0 {
```

Comment:
do you have to check if `toRemove` is filled because `corev1.Untag` is doing weird stuff when no tags are given? I think it already would be a no-op without this if, so we can remove it and improve readability?

Same below with tagging

Reply:
It does (try to) extract the object identifier. Besides that it's no-op.